src/python/pants/engine/target.py

```diff
@@ -156,34 +156,35 @@ def __repr__(self) -> str:
     def __str__(self) -> str:
         return f"{self.alias}={self.value}"
 
+    def __hash__(self) -> int:
+        return hash((self.__class__, self.value))
 
-@frozen_after_init
-@dataclass(unsafe_hash=True)
-class AsyncField(Field, metaclass=ABCMeta):
-    """A field that needs the engine in order to be hydrated.
+    def __eq__(self, other: Union[Any, Field]) -> bool:
+        if not isinstance(other, Field):
+            return NotImplemented
+        return (self.__class__, self.value) == (other.__class__, other.value)
+
+
+# NB: By subclassing `Field`, MyPy understands our type hints, and it means it doesn't matter which
+# order you use for inheriting the field template vs. the mixin.
+class AsyncFieldMixin(Field):
```

Comment:
Thanks to @cristianmatache for help with getting the mixin to work with MyPy. I couldn't get it to work via type vars or protocols, both being used to annotate the `self` for every method.

Note that this uses diamond inheritance. From Cristian:

```
           Field
         /       \
     Mixin   StringSequenceField
         \        /
          Sources
```

Cristian pointed out diamond inheritance can cause issues and be hard to follow. I'm slightly less concerned because we override `__init__()` as final, and I added lots of tests that `__eq__` and `__hash__` will work correctly with subclassing.

--

The alternative is to not subclass anything here. We have to MyPy ignore everything, and `AsyncFieldMixin` must come before the field template in the inheritance.